modules/datablocksBidAdapter.js

```diff
@@ -1,330 +1,660 @@
-import { BANNER, NATIVE, VIDEO } from '../src/mediaTypes.js';
-const NATIVE_MAP = {
-  'body': 2,
-  'body2': 10,
-  'price': 6,
-  'displayUrl': 11,
-  'cta': 12
-};
-const NATIVE_IMAGE = [{
-  id: 1,
-  required: 1,
+import { config } from '../src/config.js';
+import * as utils from '../src/utils.js';
+import { BANNER, NATIVE} from '../src/mediaTypes.js';
+import { getStorageManager } from '../src/storageManager.js';
+import {ajax} from '../src/ajax.js';
+export const storage = getStorageManager();
+
+
+const NATIVE_ID_MAP = {};
+const NATIVE_PARAMS = {
   title: {
-    len: 140
-  }
-}, {
-  id: 2,
-  required: 1,
-  img: { type: 3 }
-}, {
-  id: 3,
-  required: 1,
-  data: {
-    type: 11
-  }
-}, {
-  id: 4,
-  required: 0,
-  data: {
+    id: 1,
+    name: 'title'
+  },
+  icon: {
+    id: 2,
+    type: 1,
+    name: 'img'
+  },
+  image: {
+    id: 3,
+    type: 3,
+    name: 'img'
+  },
+  body: {
+    id: 4,
+    name: 'data',
     type: 2
+  },
+  sponsoredBy: {
+    id: 5,
+    name: 'data',
+    type: 1
+  },
+  cta: {
+    id: 6,
+    type: 12,
+    name: 'data'
+  },
+  body2: {
+    id: 7,
+    name: 'data',
+    type: 10
+  },
+  rating: {
+    id: 8,
+    name: 'data',
+    type: 3
+  },
+  likes: {
+    id: 9,
+    name: 'data',
+    type: 4
+  },
+  downloads: {
+    id: 10,
+    name: 'data',
+    type: 5
+  },
+  displayUrl: {
+    id: 11,
+    name: 'data',
+    type: 11
+  },
+  price: {
+    id: 12,
+    name: 'data',
+    type: 6
+  },
+  salePrice: {
+    id: 13,
+    name: 'data',
+    type: 7
+  },
+  address: {
+    id: 14,
+    name: 'data',
+    type: 9
+  },
+  phone: {
+    id: 15,
+    name: 'data',
+    type: 8
   }
-}, {
-  id: 5,
-  required: 0,
-  img: { type: 1 }
-}, {
-  id: 6,
-  required: 0,
-  data: {
-    type: 12
-  }
-}];
+};
+
+Object.keys(NATIVE_PARAMS).forEach((key) => {
+  NATIVE_ID_MAP[NATIVE_PARAMS[key].id] = key;
+});
 
-const VIDEO_PARAMS = ['mimes', 'minduration', 'maxduration', 'protocols', 'w', 'h', 'startdelay',
-  'placement', 'linearity', 'skip', 'skipmin', 'skipafter', 'sequence', 'battr', 'maxextended',
-  'minbitrate', 'maxbitrate', 'boxingallowed', 'playbackmethod', 'playbackend', 'delivery',
-  'pos', 'companionad', 'api', 'companiontype', 'ext'];
 
+// DEFINE THE PREBID BIDDER SPEC
 export const spec = {
-  supportedMediaTypes: [BANNER, NATIVE, VIDEO],
-  code: 'datablocks',
-  isBidRequestValid: function(bid) {
-    return !!(bid.params.host && bid.params.sourceId &&
-      bid.mediaTypes && (bid.mediaTypes.banner || bid.mediaTypes.native || bid.mediaTypes.video));
-  },
-  buildRequests: function(validBidRequests, bidderRequest) {
-    if (!validBidRequests.length) { return []; }
+	supportedMediaTypes: [BANNER, NATIVE],
+	code: 'datablocks',
+	
+	// DATABLOCKS SCOPED OBJECT
+	db_obj: {metrics_host: 'prebid.datablocks.net', metrics: [], metrics_timer: null, metrics_queue_time: 1000, vis_optout: false, source_id: 0},
```

Comment:
@htang555 isn't this better suited for your analytics adapter?

Reply:
This is for preliminary viewability metrics that are required for our adapter to perform (there is an option for the user to opt out of this in the bid params).